builder/llpcBuilderImpl.h

```diff
@@ -128,6 +133,178 @@ class BuilderImplDesc : virtual public BuilderImplBase
+                                         uint32_t                     flags,
+                                         llvm::Value*                 pCoord,
+                                         llvm::Value*                 pImageDesc,
+                                         llvm::Value*                 pSamplerDesc,
+                                         llvm::ArrayRef<llvm::Value*> address,
+                                         const llvm::Twine&           instName,
+                                         bool                         isSample);
+
+    // Common code for CreateImageAtomic and CreateImageAtomicCompareSwap
+    llvm::Value* CreateImageAtomicCommon(uint32_t                atomicOp,
+                                         uint32_t                dim,
+                                         uint32_t                flags,
+                                         llvm::AtomicOrdering    ordering,
+                                         llvm::Value*            pImageDesc,
+                                         llvm::Value*            pCoord,
+                                         llvm::Value*            pInputValue,
+                                         llvm::Value*            pComparatorValue,
+                                         const llvm::Twine&      instName);
+
+    // Change 1D or 1DArray dimension to 2D or 2DArray if needed as a workaround on GFX9+
+    uint32_t Change1DTo2DIfNeeded(uint32_t dim);
+
+    // Prepare coordinate and explicit derivatives, pushing the separate components into the supplied vectors, and
+    // modifying if necessary.
+    // Returns possibly modified image dimension.
+    uint32_t PrepareCoordinate(uint32_t                              dim,
+                               llvm::Value*                          pCoord,
+                               llvm::Value*                          pProjective,
+                               llvm::Value*                          pDerivativeX,
+                               llvm::Value*                          pDerivativeY,
+                               llvm::SmallVectorImpl<llvm::Value*>&  outCoords,
+                               llvm::SmallVectorImpl<llvm::Value*>&  outDerivatives);
+
+    // For a cubearray with integer coordinates, combine the face and slice into a single component.
+    void CombineCubeArrayFaceAndSlice(llvm::Value* pCoord, llvm::SmallVectorImpl<llvm::Value*>& coords);
+
+    // Patch descriptor with cube dimension for image call
+    llvm::Value* PatchCubeDescriptor(llvm::Value* pDesc, uint32_t dim);
+
+    // Handle cases where we need to add the FragCoord x,y to the coordinate, and use ViewIndex as the z coordinate.
+    llvm::Value* HandleFragCoordViewIndex(llvm::Value* pCoord, uint32_t flags);
+
+    // -----------------------------------------------------------------------------------------------------------------
+
+    enum ImgDataFormat
```

Comment:
Can you use exist enum ColorDataFormat or BufDataFormat?

Reply:
Probably a good idea. But this part of the diff is in the earlier PR that is now merged. So I will have to do this in a separate commit.